### DISCUSSION TITLE
A query that was dehydrated as pending ended up rejecting.

### DISCUSSION THREAD

**Discussion by SaadFarooq-Dev** [TanStack/query#8138] [Q&A] [answered]
When ever I get a error from the server I get this in the development.  Not sure what I am doing wrong here.

```js
hydration-error-info.js:63 A query that was dehydrated as pending ended up rejecting. ${QUERY_KEY}: AxiosError: Request failed with status code 500; The error will be redacted in production builds
```
This is my query
```js

export const useQueryFuncName = ({
  interval,
  enabled = false,
}: Props) => {
  return useQuery({
    enabled,
    queryKey: [Key],
    queryFn: () => {
      return func();
    },
    staleTime: 60 * 60 * 1000 * 4, // 4 hours
  });
};

```
This is the function

```js
const func = async ({ interval }: Props) => {
  const { data }: { data: TYPE } = await Api.get(
    `/path`,
    {
      params: { interval },
    },
  );
  return data;
};

```
And then we proxy to our server.

**Comment by TkDodo** [comment]
This can’t be the whole code because this error will only show up if you’re using hydration and stream a promise to the client. if you would like me to look into this, please show a minimal stackblitz or codesandbox reproduction

**Reply by SaadFarooq-Dev** [reply]
Will do. Thanks.

**Reply by SaadFarooq-Dev** [reply]
https://codesandbox.io/p/devbox/clg44m?migrateFrom=d3dtxf

![CleanShot 2024-10-08 at 01 11 45](https://github.com/user-attachments/assets/eea98f4d-6c78-45a2-b831-e6815a19be06)

**Reply by SaadFarooq-Dev** [reply]
@TkDodo

**Reply by TkDodo** [reply, marked as answer]
the problem is that your custom `dehydrateOptions` also persist erroneous / pending queries. That not only doesn’t work because you can’t serialize `Error` classes, but it also wants to store pending queries, which puts promises into the cache. Persistence and server-side-rendering use the same hydration approach, so getting promises from the persister back doesn’t really work.

What you want is to only persist successful queries, which we do per default, so:

```
import { defaultShouldDehydrateQuery } from "@tanstack/react-query";

dehydrateOptions: {
  shouldDehydrateQuery: (query) => {
    return (
      defaultShouldDehydrateQuery(query) &&
      !doNotPersistQueries.includes(query.queryKey[0])
    );
  },
}

**Reply by SaadFarooq-Dev** [reply]
Thank you @TkDodo .

**Reply by Svish** [reply]
@TkDodo Sorry to bring this back up, but I've started getting this error too and we're not even using any custom `dehydrateOptions` at all. Any idea what could cause it? Has the defaults changed or something?

The error we're getting is `A query that was dehydrated as pending ended up rejecting. [[ ... ]]: Error: CancelledError; The error will be redacted in production builds`, which also seem a bit weird as that doesn't really feel like it should be an error to begin with. 🤔

**Reply by TkDodo** [reply]
Hm, I’d need to ask @Ephem. I think the log warning could’ve sneaked in for all cases. Per default, we only dehydrate successful queries:

https://github.com/TanStack/query/blob/4e1c4338a72f7384600bbda99e44bc1891695df4/packages/query-core/src/hydration.ts#L123-L125

looking at the `dehydrateQuery` code now, the promise only gets returned if we are `pending`, so it shouldn’t be “sent”:

https://github.com/TanStack/query/blob/4e1c4338a72f7384600bbda99e44bc1891695df4/packages/query-core/src/hydration.ts#L112-L114

However, the promise is always _created_, which means the `console.error` logging might still happen:

https://github.com/TanStack/query/blob/4e1c4338a72f7384600bbda99e44bc1891695df4/packages/query-core/src/hydration.ts#L82-L94

@Ephem I think we should only create the promise if the query is in `pending` state, otherwise, we’re doing this unnecessarily ?

The situation is likely that the query is in `success` state but also has a `promise` when doing a background refetch, and if that fails, I guess you’d see the log. @Svish would be good if you could confirm this hypothesis.

**Reply by Svish** [reply]
Not sure, but that doesn't sound unreasonable at least. If I cleared out the cache in storage, then the logs did not seem to happen immediately on the next page load.

I have removed all our usage of the AbortController signal thing now, and that seems to have removed the error logging completely, even with background refetches happening. Not sure if that confirms your hypothesis. I guess if the promise created is related to the abortion then yes, but if it's related to background refetches, maybe not? 🤔

**Reply by TkDodo** [reply]
@Svish please try out the preview build of this PR:
- https://github.com/TanStack/query/pull/9847
